sdk/core/core-rest-pipeline/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -219,3 +317,74 @@ export function bearerTokenAuthenticationPolicy(
     },
   };
 }
+
+/**
+ *
+ * Interface to represent a parsed challenge.
+ *
+ * @internal
+ */
+interface AuthChallenge {
+  scheme: string;
+  params: Record<string, string>;
+}
+
+/**
+ * Converts: `Bearer a="b", c="d", Pop e="f", g="h"`.
+ * Into: `[ { scheme: 'Bearer', params: { a: 'b', c: 'd' } }, { scheme: 'Pop', params: { e: 'f', g: 'h' } } ]`.
+ *
+ * @internal
+ */
+export function parseChallenges(challenges: string): AuthChallenge[] {
+  // Split the input string by schemes and parameters
+  // Challenge regex seperates the string to individual challenges with different schemes in the format `Scheme a="b", c="d"`
+  const challengeRegex = /(\w+) ((?:\w+="[^"]*",?\s*)+)/g;
+  // Parameter regex captures the claims group removed from the scheme in the format `a="b"` and `c="d"`
+  const paramRegex = /(\w+)="([^"]*)"/g;
+
+  const parsedChallenges: AuthChallenge[] = [];
+  let match;
+
+  // Iterate over each challenge match
+  while ((match = challengeRegex.exec(challenges)) !== null) {
+    const scheme = match[1];
+    const paramsString = match[2];
+    const params: Record<string, string> = {};
+    let paramMatch;
+
+    // Iterate over each parameter match
+    while ((paramMatch = paramRegex.exec(paramsString)) !== null) {
+      params[paramMatch[1]] = paramMatch[2];
+    }
+
+    parsedChallenges.push({ scheme, params });
+  }
+  return parsedChallenges;
+}
+
+/**
+ * Parse a pipeline response and look for a CAE challenge with "Bearer" scheme
+ * Return the value in the header without parsing the challenge
+ * @internal
+ */
+function getCaeChallengeClaims(response: PipelineResponse): string | undefined {
+  const challenges = response.headers.get("WWW-Authenticate");
+  // Find all challenges present in the header
+  const parsedChallenges = parseChallenges(challenges as string);
```

Comment:
This cast is suspicious. If `getCaeChallengeClaims` was called with a response that didn't have a WWW-Authenticate header, `challenges` might be undefined and `parseChallenges` could throw a TypeError.

Could we just pass the value of the WWW-Authenticate header into this function instead of the whole response? That's the only part of the response that this helper appears to actually use.

Reply:
What happens if there are multiple WWW-Authenticate headers like the one which Timo shared? Is that valid scenario? https://datatracker.ietf.org/doc/html/rfc7616#section-3.9.1

Reply:
@KarishmaGhiya that is a valid scenario. NodeJS will be joined those headers together with comma, which is what our parsing function tries to separate out. I found [this](https://github.com/nodejs/node/blob/f67e45e0702baef1ee1253042e48fcb7b806d3b3/test/parallel/test-http-server-multiheaders.js#L34) and [this](https://github.com/nodejs/node/blob/f67e45e0702baef1ee1253042e48fcb7b806d3b3/test/parallel/test-http2-multiheaders.js#L14-L34) test for this case

Reply:
Some docs I found on the behavior:

For Node: the `WWW-Authenticate` headers will be joined by `, ` according to this doc: https://nodejs.org/api/http.html#messageheaders

For the browser, it sounds like getting the headers from a fetch response will run this algorithm: https://fetch.spec.whatwg.org/#terminology-headers

> To get a [header name](https://fetch.spec.whatwg.org/#header-name) name from a [header list](https://fetch.spec.whatwg.org/#concept-header-list) list, run these steps. They return null or a [header value](https://fetch.spec.whatwg.org/#header-value).
> 1. If list [does not contain](https://fetch.spec.whatwg.org/#header-list-contains) name, then return null.
> 1. Return the [values](https://fetch.spec.whatwg.org/#concept-header-value) of all [headers](https://fetch.spec.whatwg.org/#concept-header) in list whose [name](https://fetch.spec.whatwg.org/#concept-header-name) is a [byte-case-insensitive](https://infra.spec.whatwg.org/#byte-case-insensitive) match for name, separated from each other by 0x2C 0x20, in order.

...which is to say that the headers will be joined by `, ` (0x2c 0x20). Same thing.